### PR TITLE
Issue 933: Add support for PEM Key file formats

### PR DESCRIPTION
**Pull request by kishorekasi**
(@bug W-4203319@) TLS authentication to support PEM format X.509 
certificates along with JKS and PKCS12 formats.

Added another KeyStore Type to the list of formats (JKS, PKCS12)
currently supported during SSLContext creation.

@rev ayegorov@

Signed-off-by: Kishore Kasi Udayashankar <kudayashankar@salesforce.com>

Descriptions of the changes in this PR:

(PR description content here)...

Master Issue: #933

### CODE REVIEW

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java [hunk 1]
```diff
@@ -1414,118 +1414,137 @@ public String getClientRole() {
...
 
     /**
-     * Get the keystore type for client. Default is JKS.
+     * Get the keyfile type for client. Default is JKS.
      *
      * @return
      */
-    public String getTLSKeyStoreType() {
-        return getString(TLS_KEYSTORE_TYPE, "JKS");
+    public String getTLSKeyFileType() {
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java]: I think we cannot change all of these names it is public API.
I would prefer to keep old names

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/tls/TLSContextFactory.java [hunk 2]
```diff
@@ -70,11 +80,14 @@ private String getPasswordFromFile(String path) throws IOException {
...
         value = "OBL_UNSATISFIED_OBLIGATION",
         justification = "work around for java 9: https://github.com/spotbugs/spotbugs/issues/493")
-    private KeyStore loadKeyStore(String keyStoreType, String keyStoreLocation, String keyStorePassword)
+    private KeyStore loadKeyFile(String keyFileType, String keyFileLocation, String keyFilePassword)
             throws KeyStoreException, NoSuchAlgorithmException, CertificateException, IOException {
-        KeyStore ks = KeyStore.getInstance(keyStoreType);
-        try (FileInputStream ksin = new FileInputStream(keyStoreLocation)) {
-            ks.load(ksin, keyStorePassword.trim().toCharArray());
+        KeyStore ks = KeyStore.getInstance(keyFileType);
+        FileInputStream ksin = new FileInputStream(keyFileLocation);
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/tls/TLSContextFactory.java]: Nit: use try-with-resources

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/tls/TLSContextFactory.java [hunk 3]
```diff
@@ -282,10 +389,12 @@ public SslHandler newTLSHandler() {
         if (protocols != null && protocols.length != 0) {
             sslHandler.engine().setEnabledProtocols(protocols);
         }
+        LOG.info("Enabled cipher protocols: {} ", Arrays.toString(sslHandler.engine().getEnabledProtocols()));
```
Comment (eolivelli) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/tls/TLSContextFactory.java]: Do you need this at info level?

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/tls/TestTLS.java [hunk 4]
```diff
@@ -82,30 +105,113 @@ public void setUp() throws Exception {
...
-        baseConf.setTLSTrustStoreType("JKS");
-        baseConf.setTLSTrustStore(
-            this.getClass().getClassLoader().getResource("cacerts").toURI().getPath());
-        baseConf.setTLSTrustStorePasswordPath(
-            this.getClass().getClassLoader().getResource("trustStorePassword.txt").toURI().getPath());
+
+        switch (serverKeyFileFormat) {
+        case PEM:
+            baseConf.setTLSKeyFileType("PEM");
+            baseConf.setTLSKeyFilePath(this.getClass().getClassLoader().getResource("server-key.pem").getPath());
```
Comment (eolivelli) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/tls/TestTLS.java]: Do you think it is possible to generate these files at build time instead of using static files?
Certificates will expire one day.
If not so simple I am fine with keeping files, but having a readme files which explains how to rebuild the files in each format

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/tls/TLSContextFactory.java (hunk 2 above)
Comment (kishorekasi) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/tls/TLSContextFactory.java]: ok

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java (hunk 1 above)
Comment (kishorekasi) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java]: will revert to old names

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/tls/TLSContextFactory.java (hunk 3 above)
Comment (kishorekasi) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/tls/TLSContextFactory.java]: Not necessarily; this can move to debug

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java [hunk 5]
```diff
@@ -145,14 +145,6 @@
     // Client auth provider factory class name. It must be configured on Bookies to for the Auditor
     protected static final String CLIENT_AUTH_PROVIDER_FACTORY_CLASS = "clientAuthProviderFactoryClass";
 
-    // Client TLS
-    protected static final String TLS_KEYSTORE_TYPE = "clientKeyStoreType";
```
Comment (sijie) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java]: this breaks BC. 

for each field you are change here, I think you have to do:

- in this file: change "TLS_XYZ" to "CLIENT_XYZ", but not change the value.
- change getTLSxyz() to : getProperty(TLS_XYZ, getProperty(CLIENT_XYZ, ...))

I am not what AutoRecovery is using, is it using client settings or server settings? if it is using client settings, you need also to update the configuration settings in the bookie conf.

Comment (kishorekasi) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java]: I am moving TLS_XYZ from ServerConfiguration and ClientConfiguration to the common class AbstractConfiguration and just use one set of config params (no clientXYZ params anymore). So, effectively this is a dead code. What I am doing here is to keep just one set of config params so both bookie and client config params names are same.

AutoRecovery can still use the same set of params as bookies to configure the bookkeeper client.

Only config param names are changing, but the variable name still remains the same as TLS_XYZ. Do you still think its a BC issue?

Comment (sijie) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/conf/ClientConfiguration.java]: yeah param name changing is considered as breaking BC. because if a client upgrades from 4.6.0 to 4.7.0, without changing the configuration, it won't work.

**Comment by sijie** [comment]
fyi - had a conversation with @kishorekasi offline. we agreed on the approach for addressing the BC concerns here. @kishorekasi will update this PR.

The example PR on showing how to deal with the BC concerns is here: https://github.com/kishorekasi/bookkeeper/pull/1

File: bookkeeper-server/pom.xml [hunk 6]
```diff
@@ -284,16 +284,37 @@
...
+        <executions>
+            <execution>
+                <id>Generate Self-Signed Certificates</id>
+                <phase>generate-test-resources</phase>
+                <goals>
+                    <goal>exec</goal>
+                </goals>
+                <configuration>
+                    <workingDirectory>${basedir}/src/test/resources</workingDirectory>
+                    <executable>${basedir}/src/test/resources/generateKeysAndCerts.sh</executable>
```
Comment (eolivelli) [inline bookkeeper-server/pom.xml]: This is going to create new certificates at every build, two problems:
1) this is slow
2) it requires some jind of setup of openssl

I suggest to add a profile to build the certificates and commit a working version of certificates.
When they will expire we will do another commit

Comment (sijie) [reply bookkeeper-server/pom.xml]: How slow is this?

**Comment by kishorekasi** [comment]
2-3 seconds on my Ubuntu server

File: bookkeeper-server/pom.xml (hunk 6 above)
Comment (eolivelli) [reply bookkeeper-server/pom.xml]: With this patch at every "mvn test" we are going to re-create new certificates.
I think we have two options:
1) commit certificates and rebuild them when expired
2) make the script rebuild the certificate only if it does not exist

I prefer 1)as it makes tests reproducible at 100% and it does not impact build and dev time

Comment (kishorekasi) [reply bookkeeper-server/pom.xml]: @eolivelli If it makes sense, I can commit certificates with a very long expiry so that we may never need to generate the certificates again. Also move certification generation to a new profile, just in case we need it.

Comment (eolivelli) [reply bookkeeper-server/pom.xml]: It would be fine for me, thank you

Comment (sijie) [reply bookkeeper-server/pom.xml]: +1 on @kishorekasi 's proposed approach.
